File: src/client/ts/model/variant.ts
import { JSONObject } from 'harmony-types';
import { OptionJSON } from '../responses/option';
import { Options } from './options';

export class Variant {
	#id = '';
	#name = '';
	#thumbnailUrl = '';
	#retailPrice = 0;
	#currency = '';
	#options = new Options();

	get id(): string {
		return this.#id;
	}

	set id(id) {
		this.#id = id;
	}

	get name(): string {
		return this.#name;
	}

	set name(name) {
		this.#name = name;
	}

	get thumbnailUrl(): string {
		return this.#thumbnailUrl;
	}

	set thumbnailUrl(thumbnailUrl) {
		this.#thumbnailUrl = thumbnailUrl;
	}

	get retailPrice(): number {
		return this.#retailPrice;
	}

	set retailPrice(retailPrice) {
		this.#retailPrice = Number(retailPrice);
	}

	get currency(): string {
		return this.#currency;
	}

	set currency(currency) {
		this.#currency = currency;
	}

	get options(): Options {
		return this.#options;
	}

	set options(options) {
		this.#options = options;
	}

	fromJSON(shopProductJson: JSONObject = {}): void {
		this.id = shopProductJson.id as string;
		this.name = shopProductJson.name as string;
		this.thumbnailUrl = shopProductJson.thumbnail_url as string;
		this.retailPrice = shopProductJson.retail_price as number;
		this.#currency = shopProductJson.currency as string;
		this.#options.fromJSON(shopProductJson.options as OptionJSON[]);
	}

	toJSON() {
		return {
			id: this.id,
			name: this.name,
			thumbnail_url: this.thumbnailUrl,
			retail_price: this.retailPrice,
			currency: this.#currency,
			options: this.#options.toJSON(),
		};
	}
}

File: src/server/model/variant.go
package model

import (
	"github.com/shopspring/decimal"
)

type Variant struct {
	ID           string          `json:"id" bson:"id"`
	Name         string          `json:"name" bson:"name"`
	ThumbnailURL string          `json:"thumbnail_url" bson:"thumbnail_url"`
	RetailPrice  decimal.Decimal `json:"retail_price" bson:"retail_price"`
	Currency     string          `json:"currency" bson:"currency"`
	Options      []Option        `json:"options" bson:"options"`
}

func NewVariant(product *Product) Variant {
	return Variant{
		ID:           product.ID,
		Name:         product.Name,
		ThumbnailURL: product.ThumbnailURL,
		//RetailPrice:  product.RetailPrice,
		//Currency: product.Currency,
		Options: product.Options,
	}
}
